### PR TITLE
Parse and store HullType with unique_ptr

### LINKED ISSUE #1548: ShipHullParser should return unique_ptr and not leak memory.
**Environment: all**

**Description**
The `parse/ShipHullParser.cpp` is a generator.  In C++11 a generator should return a `std::unique_ptr<>` to indicate that it has created something that requires memory management and that it is passing management responsibility to the calling function.  The calling function can chose to receive the pointer as a `unique_ptr<>` or transfer it to a `shared_ptr<>` if it needs a shared ownership model.

Furthermore the error condition near line 38 leaks the passed in pointer.  This is a static leak, it only leaks objects that have failed to parse once per game.

**Expected Result**
Return `std::unique_ptr<>`s.

**Steps to reproduce**
Play game.

**Suggestions**
A solution of the same issue in the BuildingType parser is in #1544, is anyone would like to assign themselves and help out.

### PR DESCRIPTION
**Pull request by dbenage-cx**
Resolves #1548

### CODE REVIEW

File: universe/ShipDesign.cpp [hunk 1]
```diff
@@ -650,11 +650,13 @@ HullTypeManager::HullTypeManager() {
         throw;
     }
 
-    TraceLogger() << "Hull Types:";
-    for (const auto& entry : m_hulls) {
-        const HullType* h = entry.second;
-        TraceLogger() << " ... " << h->Name();
-    }
+    TraceLogger() << [&]() {
+            std::string retval("Hull Types:");
```
Comment (LGM-Doyle) [inline universe/ShipDesign.cpp]: I prefer explicit capture of `[this]` here.

In this location it make no difference.

Implicit capture as a practice throughout the code base, may easily lead to capturing a variable with shorter scope than the lambda.  Given that the code  is written by many hands, there are many differing uses of scope and commonly using unnamed capture will make it harder to debug crashes due to a lambda referencing an out of scope object.

**Review by LGM-Doyle — COMMENTED** [review]
Since there will be minor conflicts, renaming `X*` to `uniqueptr<X>` with #1785, I can wait for you to finish all of these bug fixes before merging any of #1785, so that I do the rebasing work.

**Comment by dbenage-cx** [comment]
> all of these bug fixes

@LGM-Doyle Do you refer to both PRs, or is there more than one in this PR?
Lately, github notices are not always showing, so want to make sure that is not extending to review comments.

**Comment by LGM-Doyle** [comment]
You have changed everything I requested.

You have fixed 3 of the 6 memory leaks in the parser.  I was hoping that you would fix the other 3 as well.  I was suggesting that if you plan to fix the other 3, that I can wait to merge the parts of #1785 that would conflict with the fixes until after you have completed the fixes.  The conflicts would  be minor, but the offer is there.

**Comment by dbenage-cx** [comment]
I need to familiarize myself with valgrind and identify those leaks.
Please do not feel a need to wait on me, especially as it may be easier to rebase these smaller PRs.

**Pull request merged**